### PR TITLE
fxswap-ext-fee

### PR DESCRIPTION
**Pull request by heswithme**
This PR contains FXSwap V3 pools.

Main changes:
- external policy hook for fee and price-scale guidance
- policy-state synchronization from pool state
- LP/admin profit accounting hardening
- donation/rebalancing edge-case fixes
- additional guards and tests around ramping, fees, and withdrawal paths

### CODE REVIEW

File: contracts/main/StableswapMath.vy [hunk 1]
```diff
@@ -74,6 +74,11 @@ def newton_D(_amp: uint256,
     """
     # gamma and K0_prev are ignored
     # _amp is already multiplied by a A_MULTIPLIER and N_COINS
+    assert _xp[0] > 0 and _xp[1] > 0, "!balance"
+    if _xp[0] > _xp[1]:
+        assert unsafe_div(_xp[0], _xp[1]) < 10_000, "!balance"
+    else:
+        assert unsafe_div(_xp[1], _xp[0]) < 10_000, "!balance"
```
Comment (michwill) [inline contracts/main/StableswapMath.vy]: NICE!

File: contracts/main/Twocrypto.vy [hunk 2]
```diff
@@ -111,8 +121,8 @@ event NewParameters:
     mid_fee: uint256
     out_fee: uint256
     fee_gamma: uint256
-    allowed_extra_profit: uint256
-    adjustment_step: uint256
+    adjustment_step_min: uint256
```
Comment (michwill) [inline contracts/main/Twocrypto.vy:131]: You still want to include this guy just for gas savings?

File: contracts/main/Twocrypto.vy [hunk 3]
```diff
@@ -49,6 +49,16 @@ interface Views:
         i: uint256, j: uint256, dy: uint256, swap: address, n_iter: uint256
     ) -> uint256: view
 
+interface Policy:
+    def get_fee(xp: uint256[N_COINS], packed_fee_params: uint256) -> uint256: view
```
Comment (michwill) [inline contracts/main/Twocrypto.vy]: We may want to include xp before and after the swap (the second one can be zeros if it's not a swap). Could be good for making the True Dynamic Spread. Not sure how complex that is to do tho

File: contracts/main/Twocrypto.vy [hunk 4]
```diff
@@ -195,16 +212,19 @@ virtual_price: public(uint256)  # <------ Cached (fast to read) virtual price.
...
+# Split between rebalancing budget and admin/LP share
+lp_profit_fraction: public(uint256)
+
 admin_fee: public(uint256)
-MAX_ADMIN_FEE: constant(uint256) = 10**10
-MIN_FEE: constant(uint256) = 5 * 10**5  # <-------------------------- 0.5 BPS.
-MAX_FEE: constant(uint256) = 10 * 10**9
-NOISE_FEE: constant(uint256) = 10**5  # <---------------------------- 0.1 BPS.
+MAX_ADMIN_FEE: constant(uint256) = FEE_PRECISION
+MIN_FEE: constant(uint256) = FEE_PRECISION * 1 // 2 // 10_000  # <-------------------------- 0.5 BPS.
```
Comment (michwill) [inline contracts/main/Twocrypto.vy]: Btw I wonder if we want it to be 0.5 bps or even smaller

File: contracts/main/Twocrypto.vy [hunk 5]
```diff
@@ -254,9 +287,13 @@ def __init__(
...
     self.VIEW = Views(empty(address))
     self.MATH = Math(empty(address))
+    self.POLICY = Policy(empty(address))
 
     # this parameter can also be dynamically adjusted at blueprint deployment time
-    self.admin_fee = 10**10 * 50 // 100
+    self.admin_fee = FEE_PRECISION * 50 // 100
+
+    # Split between rebalancing budget and admin/LP share
+    self.lp_profit_fraction = PRECISION * 50 // 100
```
Comment (michwill) [inline contracts/main/Twocrypto.vy]: We will be able to set in init right?

File: contracts/main/Twocrypto.vy [hunk 6]
```diff
@@ -1051,8 +1088,8 @@ def tweak_price(
         # ----------------- We cap state price that goes into the EMA with
         #                                                 2 x price_scale.
         price_oracle = unsafe_div(
-            min(last_prices, 2 * price_scale) * (10**18 - alpha) +
-            price_oracle * alpha,  # ^-------- Cap spot price into EMA.
+            min(max(last_prices, price_scale // 2), 2 * price_scale) * (10**18 - alpha) +
```
Comment (michwill) [inline contracts/main/Twocrypto.vy]: btw if you will ever be out of bytespace - `unsafe_div` is ok here

File: contracts/main/Twocrypto.vy [hunk 7]
```diff
@@ -1105,20 +1143,17 @@ def tweak_price(
...
+    threshold_vp: uint256 = max(10**18, 10**18 + xcp_profit * self.lp_profit_fraction // PRECISION - self.lp_profit_fraction)
 
     # user_supply < total_supply => vp_boosted > virtual_price
     # by not accounting for donation shares, virtual_price is boosted leading to rebalance trigger
     # this is approximate condition that preliminary indicates readiness for rebalancing
     vp_boosted: uint256 = 10**18 * xcp // locked_supply
     assert vp_boosted >= virtual_price, "negative donation"
-    if (vp_boosted  > threshold_vp + rebalancing_params[0]) and (last_timestamp < block.timestamp):
-        #        allowed_extra_profit --------^               #   ^ only rebalance once per block (first tx)
+    if (vp_boosted  > threshold_vp) and (last_timestamp < block.timestamp):
```
Comment (michwill) [inline contracts/main/Twocrypto.vy]: 👍

File: contracts/main/Twocrypto.vy [hunk 8]
```diff
@@ -1020,9 +1056,10 @@ def tweak_price(
     last_prices: uint256 = self.last_prices
     price_scale: uint256 = self.cached_price_scale
     rebalancing_params: uint256[3] = self._unpack_3(self.packed_rebalancing_params)
```
Comment (michwill) [inline contracts/main/Twocrypto.vy:1116]: [crazy thought] I wonder if this may work:
```
struct RebalancingParams:
    ...

...
rebalancing_params: RebalancingParams = convert(self._unpack_3(...), RebalancingParams)
```
for readability later

File: contracts/main/Twocrypto.vy (hunk 2 above)
Comment (heswithme) [reply contracts/main/Twocrypto.vy:131]: yep, that is another knob to use to avoid dust rebalancing

File: contracts/main/Twocrypto.vy (hunk 5 above)
Comment (heswithme) [reply contracts/main/Twocrypto.vy]: factory being cooked

File: contracts/main/Twocrypto.vy (hunk 4 above)
Comment (heswithme) [reply contracts/main/Twocrypto.vy]: good q to you. also - do we want to impose this min_fee on external policy? or can external policy decide its own bounds?

File: contracts/main/StableswapMath.vy (hunk 1 above)
Comment (heswithme) [reply contracts/main/StableswapMath.vy]: 10k ok? should be unreasonable to reach for any practical matter

File: contracts/main/Twocrypto.vy (hunk 3 above)
Comment (heswithme) [reply contracts/main/Twocrypto.vy]: this xp is already post-swap, pre-fee, and policy already has pre-swap xp from historical state. so, already done?

File: contracts/main/Twocrypto.vy [hunk 9]
```diff
@@ -1421,6 +1477,11 @@ def _A_gamma() -> uint256[2]:
 @view
 def _fee(xp: uint256[N_COINS]) -> uint256:
 
+    if self.POLICY != empty(Policy):
+        fee: uint256 = staticcall self.POLICY.get_fee(xp, self.packed_fee_params)
+        assert fee <= MAX_FEE, "fee>MAX"
```
Comment (michwill) [inline contracts/main/Twocrypto.vy]: Cannot the external contract here block the smart contract? Is it a good thing?
Or is it better to `return min(fee, MAX_FEE)` instead

File: contracts/main/Twocrypto.vy (hunk 4 above)
Comment (heswithme) [reply contracts/main/Twocrypto.vy]: lowered to 0.1bps and clamped policy returned fee

File: contracts/main/Twocrypto.vy (hunk 5 above)
Comment (heswithme) [reply contracts/main/Twocrypto.vy]: revised with initialize function for tx.origin

File: contracts/main/Twocrypto.vy (hunk 6 above)
Comment (heswithme) [reply contracts/main/Twocrypto.vy]: 15% of bytecode were assert messages. cleaned up where could.

File: contracts/main/Twocrypto.vy (hunk 9 above)
Comment (heswithme) [reply contracts/main/Twocrypto.vy]: clamped

File: contracts/main/Twocrypto.vy [hunk 10]
```diff
@@ -1138,7 +1149,10 @@ def tweak_price(
     # Rebalancing condition transformation:
     # virtual_price > 1 + (xcp_profit - 1) * lp_profit_fraction
     # virtual_price > 1 + xcp_profit * lp_profit_fraction - lp_profit_fraction
-    threshold_vp: uint256 = max(10**18, 10**18 + xcp_profit * self.lp_profit_fraction // PRECISION - self.lp_profit_fraction)
+    threshold_vp: uint256 = max(
+        PRECISION,
+        PRECISION + xcp_profit * self.lp_profit_fraction // FEE_PRECISION - PRECISION * self.lp_profit_fraction // FEE_PRECISION
```
Comment (michwill) [inline contracts/main/Twocrypto.vy]: I strongly suspect that `self.lp_profit_fraction * (PRECISION // FEE_PRECISION)` will eat less bytecode!

**Comment by heswithme** [comment]
all planned features added, next stage - internal review & audit, then merge

File: contracts/main/Twocrypto.vy [hunk 11]
```diff
@@ -2206,69 +2277,166 @@ def apply_new_parameters(
...
+    self._set_policy(policy)
+
+    # Set allowlist
+    self.lp_allowlist[empty(address)] = False
+    for account: address in allowlist_add:
+        if account != empty(address):
+            self.lp_allowlist[account] = True
+            log LPAllowlistChanged(user=account, allowed=True)
+
+    if len(allowlist_add) > 0:
```
[the remark this answers is not in the document]
Comment (heswithme) [reply contracts/main/Twocrypto.vy]: good catch clanker

File: contracts/main/Twocrypto.vy [hunk 12]
```diff
@@ -268,6 +306,11 @@ def __init__(
     # --------------- Validate A and gamma parameters here and not in factory.
     gamma_A: uint256[2] = self._unpack_2(packed_gamma_A)  # gamma is at idx 0.
 
+    if gamma_A[0] == 11111111111:
+        # magic value that enables pool initialization by deployer (set whitelist, admin_fee, lpf)
+        self.deploy_eoa = tx.origin
```
[the remark this answers is not in the document]
Comment (heswithme) [reply contracts/main/Twocrypto.vy:322]: our pools are deployed via factory, so its concise choice

File: contracts/main/Twocrypto.vy [hunk 13]
```diff
@@ -2206,69 +2277,167 @@ def apply_new_parameters(
...
+    @param allowlist_add Initial LP allowlist entries. Non-empty input enables the whitelist.
     """
 
+    # Access control: only deployer during the first 4 hours after deployment, then only admin.
+    deploy_time: uint256 = self.deploy_time
+    assert deploy_time != 0  # dev: "pool does not need initialization"
+    assert self.D == 0  # dev: "pool already has liquidity"
+
+    if block.timestamp <= deploy_time + 4 * 3600: # within the first 4h after pool creation
+        assert msg.sender == self.deploy_eoa  # dev: "only deployer during initialization window"
```
[the remark this answers is not in the document]
Comment (heswithme) [reply contracts/main/Twocrypto.vy]: factory deployer.

File: contracts/main/TwocryptoView.vy [hunk 14]
```diff
@@ -282,73 +282,35 @@ def _calc_dtoken_nofee(
...
 
 
 @internal
 @view
 def _fee(xp: uint256[N_COINS], swap: address) -> uint256:
+    policy: Policy = staticcall Curve(swap).POLICY()
+    if policy != empty(Policy):
+        fee: uint256 = staticcall policy.get_fee(xp, staticcall Curve(swap).packed_fee_params())
+        if fee != 0:
+            return min(FEE_PRECISION, max(MIN_FEE, fee))
```
[the remark this answers is not in the document]
Comment (heswithme) [reply contracts/main/TwocryptoView.vy]: dead code indeed, we keep it because why not

File: contracts/main/Twocrypto.vy [hunk 15]
```diff
@@ -2006,22 +2080,12 @@ def fee_gamma() -> uint256:
...
-def adjustment_step() -> uint256:
+def adjustment_step() -> uint256[2]:
     """
-    @notice Returns the current adjustment step
-    @return uint256 adjustment_step value.
+    @notice Returns the current adjustment steps
+    @return uint256 adjustment_steps values: [min, max].
     """
-    return self._unpack_3(self.packed_rebalancing_params)[1]
+    return [self._unpack_3(self.packed_rebalancing_params)[0], self._unpack_3(self.packed_rebalancing_params)[1]]
```
Comment (romanagureev) [inline contracts/main/Twocrypto.vy]: Save and reuse?

File: contracts/main/Twocrypto.vy [hunk 16]
```diff
@@ -2162,42 +2226,42 @@ def apply_new_parameters(
...
+        new_adjustment_step_min = current_rebalancing_params[0]
 
-    if new_adjustment_step > 10**18:
-        new_adjustment_step = current_rebalancing_params[1]
+    if new_adjustment_step_max > 10**18:
+        new_adjustment_step_max = current_rebalancing_params[1]
 
     if new_ma_time < 872542:  # <----- Calculated as: 7 * 24 * 60 * 60 / ln(2)
-        assert new_ma_time > 86, "MA<60/ln(2)"
+        assert new_ma_time > 86  # dev: "MA time below minimum"
```
Comment (romanagureev) [inline contracts/main/Twocrypto.vy]: Could you remain comment about `60/ln(2)`, so it's not a magic number

File: contracts/main/Twocrypto.vy [hunk 17]
```diff
@@ -2162,42 +2226,42 @@ def apply_new_parameters(
...
 
-    if new_allowed_extra_profit > 10**18:
-        new_allowed_extra_profit = current_rebalancing_params[0]
+    if new_adjustment_step_min > 10**18:
+        new_adjustment_step_min = current_rebalancing_params[0]
 
-    if new_adjustment_step > 10**18:
-        new_adjustment_step = current_rebalancing_params[1]
+    if new_adjustment_step_max > 10**18:
+        new_adjustment_step_max = current_rebalancing_params[1]
```
Comment (romanagureev) [inline contracts/main/Twocrypto.vy:2356]: assert new min less than new max

File: contracts/main/Twocrypto.vy [hunk 18]
```diff
@@ -1269,92 +1364,58 @@ def _claim_admin_fees():
...
+        # Keep xcp_profit as the gross profit signal and track claimed admin
+        # extraction separately in the same virtual-price units as `fees`.
+        # Rebalancing thresholding then compares virtual_price against:
+        #   1 + (xcp_profit - 1) * lpf - admin_claimed_profit
+        self.admin_claimed_profit += fees
+
+        # ---------------------------- Update State ------------------------------
+        self.virtual_price = updated_vprice
+        self.last_admin_fee_claim_timestamp = block.timestamp
+        self.xcp_profit_a = xcp_profit  # <-------- Cache last claimed gross profit.
```
Comment (romanagureev) [inline contracts/main/Twocrypto.vy]: Logically `xcp_profit` should be affected

File: contracts/main/Twocrypto.vy [hunk 19]
```diff
@@ -1421,6 +1485,12 @@ def _A_gamma() -> uint256[2]:
 @view
 def _fee(xp: uint256[N_COINS]) -> uint256:
 
+    if self.POLICY != empty(Policy):
+        fee: uint256 = staticcall self.POLICY.get_fee(xp)
```
Comment (romanagureev) [inline contracts/main/Twocrypto.vy:1618]: Is it reasonable to do `raw_call` with success check?

File: contracts/main/Twocrypto.vy [hunk 20]
```diff
@@ -2206,84 +2267,173 @@ def apply_new_parameters(
...
+@internal
+def _set_allowlist(add: DynArray[address, 16], remove: DynArray[address, 16]):
+    # The allowlist enabled flag is stored at `lp_allowlist[empty(address)]`.
+    # `empty(address)` in `add` is ignored, while `empty(address)` in `remove`
+    # disables the allowlist. After processing both arrays, we always emit the
+    # final sentinel state once.
+    for account: address in remove:
+        if account == empty(address):
+            self.lp_allowlist[empty(address)] = False
+            continue
```
Comment (romanagureev) [inline contracts/main/Twocrypto.vy:2447]: Why do you need this extra case?

File: contracts/main/Twocrypto.vy [hunk 21]
```diff
@@ -2206,84 +2267,173 @@ def apply_new_parameters(
...
+        self.lp_allowlist[account] = False
+        log LPAllowlistChanged(user=account, allowed=False)
+
+    for account: address in add:
+        if account == empty(address):
+            continue
+
+        self.lp_allowlist[account] = True
+        log LPAllowlistChanged(user=account, allowed=True)
+        self.lp_allowlist[empty(address)] = True
```
Comment (romanagureev) [inline contracts/main/Twocrypto.vy:2458]: You can do `if len(add) > 0: self.lp_allowlist[empty(address)] = True`, though this can be unintuitive without dedicated turning on/off for the user anyway :(

File: contracts/main/Twocrypto.vy (hunk 17 above)
Comment (heswithme) [reply contracts/main/Twocrypto.vy:2356]: asserted, ok

File: contracts/main/Twocrypto.vy (hunk 16 above)
Comment (heswithme) [reply contracts/main/Twocrypto.vy]: fixed

File: contracts/main/Twocrypto.vy (hunk 19 above)
Comment (heswithme) [reply contracts/main/Twocrypto.vy:1618]: policy is trusted

File: contracts/main/Twocrypto.vy (hunk 21 above)
Comment (heswithme) [reply contracts/main/Twocrypto.vy:2458]: keeping for event logging semantics
